106-self-healing-feature-in-operator.md

```diff
@@ -0,0 +1,295 @@
+For example, the `GoalViolations` class uses the `RebalanceRunnable` class, the `DiskFailure` class use the `RemoveDisksRunnable` class and so on.
+An optimization proposal will then be generated by these `Runnable` classes and that proposal will be applied on the cluster to fix the anomaly.
+In case the anomaly detected is unfixable for e.g. violated hard goals that cannot be fixed typically due to lack of physical hardware (insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy Replica Capacity Goal, or insufficient number of resources to satisfy resource capacity goals), the anomaly wouldn't be fixed and the Cruise Control logs will be updated with `self-healing is not possible due to unfixable goals` warning.
+
+### Proposal
+
+This proposal allows the users to have their cluster balanced automatically whenever an anomaly is detected by Cruise Control.
+To ensure that the operator drives the rebalance we will be leveraging the Cruise Control self-healing mechanism but not completely.
+When using self-healing in Cruise Control, the rebalance are triggered automatically in the cluster which means that the operator wouldn't have the information about when a rebalance is happening.
+To resolve this issue, we will only make use of the ability of Cruise Control to detect the anomalies and based on the detection, we will then notify the operator to run the rebalance.
+Doing this will provide us with following merits:
+* ensure that the operator know what is going on in the cluster.
+* better debugging.
+* to drive the rebalances/fixes from the operator.
+
+#### `Full` mode in auto-rebalancing feature of Strimzi
+
+To leverage the above self-healing mechanism, we will be introducing a new mode to the `auto-rebalancing` feature.
+The new mode will be called `FULL`, which means that an anomaly was detected and the rebalancing should be applied to the whole cluster
+The mode is defined by setting the `spec.cruiseControl.autoRebalance.mode` field as `full` and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource (read later for more details), by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+Once the auto-rebalance with `Full` mode is enabled, self-healing will be enabled and a notifier `AnomalyDetectorNotifier` will be set(default notifier).
+This notifier's job will be to update the operator regarding the anomalies being detected so that the operator can trigger a rebalance.
+The user can change the default notifier if they want by setting the `anomaly.notifier.class` to point towards their own custom notifier.
```

Comment:
But if they are going to do that, they will miss the auto fix from Strimzi.
Also it seems you are proposing to configure our own notifier if the full auto-rebalancing is set but we should clarify that a manual user configuration would override it.